tl

```diff
@@ -245,92 +300,456 @@ for i, filename in ipairs(args["script"]) do
+   -- with pattern matching while iterating over a project
+   -- paths[files.foo.bar] -> "foo/bar"
+   local emptyref = emptyref or {}
+   for file in lfs.dir(dirname) do
+      if file ~= "." and file ~= ".." then
+         if lfs.attributes(path_concat(dirname, file), "mode") == "directory" then
+            local p
+            local prefix = dirname
+            files[file], p = traverse(path_concat(dirname, file), emptyref)
+            paths[files[file]] = file
+            for k, v in pairs(p) do
+               paths[k] = path_concat(file, v)
+            end
+         else
+            -- storing a special entry in this table to it mark as empty could
+            -- interfere with convoluted or maliciously constructed directory
+            -- names so we use a table with specific metatable to mark
+            -- something as the end of a traversal to have a property attached
+            -- to the table, without creating an entry in the table
+            files[file] = setmetatable({}, emptyref)
+            paths[files[file]] = file
+         end
       end
-      -- put script in arg[0] and arguments in positive positions
-      for p, a in ipairs(args["script"]) do
-         arg[p - 1] = a
+   end
+   return files, paths, emptyref
+end
+
+local project = {} -- This will probably get exposed in the api if that happens
+function project:files(inc_patt_arr, exc_patt_arr) -- iterate over the files in the project adhering to the provided patterns
+   inc_patt_arr = inc_patt_arr or {}
+   exc_patt_arr = exc_patt_arr or {}
+   local function iter(dirs, prefix)
+      prefix = prefix or ""
+      for fname, file in pairs(dirs) do
+         local path = self.paths[file]
+         if getmetatable(file) == self.emptyref then
+            local include = true
+
+            if tlconfig["files"] then
+               include = false
+            end
+            if #inc_patt_arr > 0 then
+               local idx = inc_patt_arr:match(path)
+               if not idx then
+                  include = false
+               end
+            end
+            if #exc_patt_arr > 0 then
+               idx = exc_patt_arr:match(path)
+               if include and idx then
+                  include = false
+               end
+            end
+            if include then
+               coroutine.yield(self.paths[file])
+            end
+         else
+            iter(file, fname)
+         end
       end
-      -- cleanup the rest
-      n = nargs
-      while arg[n] do
-         arg[n] = nil
-         n = n + 1
+   end
+   return coroutine.wrap(iter), self.dir
+end
+function project:find(path) -- allow for indexing with paths project:find("foo/bar") -> project.dir.foo.bar
+   local current_dir = self.dir
+   for dirname in str_split(path, path_separator) do
+      current_dir = current_dir[dirname]
+      if not current_dir then
+         return nil
       end
+   end
+   return current_dir
+end
+local function is_in(a, b) -- a is in b
```

Comment:
looks like this function is unused

Reply:
I think I had a plan at one point to print warnings when generating files in the `source_dir` when `build_dir` was present. :shrug: